coral/src/app/features/topics/request/schemas/topic-request-form.ts

```diff
@@ -0,0 +1,213 @@
+import z, { RefinementCtx } from "zod";
+import isNumber from "lodash/isNumber";
+import { UseFormReturn } from "react-hook-form";
+import { useEffect } from "react";
+
+const topicNameField = z
+  .string()
+  .min(1, { message: "Topic name can not be empty" })
+  .max(255, { message: "Topic name length can not exceed 255 characters" });
+
+const topicPartitionsField = z.string();
+
+const replicationFactorField = z.string();
+
+const environmentField = z.object({
+  name: z.string(),
+  id: z.string(),
+  maxReplicationFactor: z.number().optional(),
+  maxPartitions: z.number().optional(),
+  defaultPartitions: z.number().optional(),
+  defaultReplicationFactor: z.number().optional(),
+  topicNamePrefix: z.string().optional(),
+  topicNameSuffix: z.string().optional(),
+});
+
+const formSchema = z
+  .object({
+    environment: environmentField,
+    topicpartitions: topicPartitionsField,
+    replicationfactor: replicationFactorField,
+    topicname: topicNameField,
+    remarks: z.string(),
+    description: z.string(),
+  })
+  .superRefine(validateTopicPartitions)
+  .superRefine(validateReplicationFactor)
+  .superRefine(validateTopicName);
+
+function validateReplicationFactor(
+  val: {
+    environment: z.infer<typeof environmentField>;
+    replicationfactor: z.infer<typeof replicationFactorField>;
+  },
+  ctx: RefinementCtx
+) {
+  const {
+    environment: { maxReplicationFactor },
+    replicationfactor,
+  } = val;
+  if (
+    isNumber(maxReplicationFactor) &&
+    parseInt(replicationfactor, 10) > maxReplicationFactor
+  ) {
+    ctx.addIssue({
+      code: z.ZodIssueCode.too_big,
+      inclusive: true,
+      maximum: maxReplicationFactor,
+      type: "number",
+      path: ["replicationfactor"],
+      message: `${replicationfactor} can not be bigger than ${maxReplicationFactor}`,
+    });
+  }
+}
+
+function validateTopicPartitions(
+  val: {
+    environment: z.infer<typeof environmentField>;
+    topicpartitions: z.infer<typeof topicPartitionsField>;
+  },
+  ctx: RefinementCtx
+) {
+  const {
+    environment: { maxPartitions },
+    topicpartitions,
+  } = val;
+  if (
+    isNumber(maxPartitions) &&
+    parseInt(topicpartitions, 10) > maxPartitions
+  ) {
+    ctx.addIssue({
+      code: z.ZodIssueCode.too_big,
+      inclusive: true,
+      maximum: maxPartitions,
+      type: "number",
+      path: ["topicpartitions"],
+      message: `${topicpartitions} can not be bigger than ${maxPartitions}`,
+    });
+  }
+}
+
+function validateTopicName(
+  val: {
+    environment: z.infer<typeof environmentField>;
+    topicname: z.infer<typeof topicNameField>;
+  },
+  ctx: RefinementCtx
+) {
+  const {
+    environment: { topicNamePrefix, topicNameSuffix },
+    topicname,
+  } = val;
+  if (topicNamePrefix !== undefined && !topicname.startsWith(topicNamePrefix)) {
+    ctx.addIssue({
+      code: z.ZodIssueCode.custom,
+      fatal: true,
+      message: `Topic name must start with "${topicNamePrefix}".`,
```

Comment:
Noticed that the validation will allow the topic to be named as *only* the prefix. Is this valid/expected, or should we also enforce at least prefix + 1 character ?

Also, should we validate name collision on topic name at this validation stage? If name collision is a concern.

https://user-images.githubusercontent.com/20607294/211306724-4c2e13d9-de84-40d7-87bb-5b3848def6aa.mov

Reply:
> Noticed that the validation will allow the topic to be named as only the prefix. Is this valid/expected, or should we also enforce at least prefix + 1 character ?

I expect the prefix is still a valid topic name. 

> Also, should we validate name collision on topic name at this validation stage? If name collision is a concern.

I would leave it to the API. The validation will happen there anyways and not sure if we want to load all topic in this view in order to make a comparison whether a name is unique or not

Reply:
I think relying on backend validation for this is good. However, we do already fetch the topic names elsewhere in the app (ACL request), so maybe it would not be too expensive to do it here as well (there might be already cached values).

But it's probably on the "make it nice" list which is not our priority haha!